src/tools/compiletest/src/runtest.rs

```diff
@@ -992,7 +997,7 @@ actual:\n\
     fn check_expected_errors(&self,
                              expected_errors: Vec<errors::Error>,
                              proc_res: &ProcRes) {
-        if proc_res.status.success() {
+        if proc_res.status.success() && expected_errors.iter().any(|x| x.kind == Some(ErrorKind::Error)) {
```

Comment:
This change is technically correct, in that `check_expected_errors` is only called from one place before the change, `run_cfail_test` and it checks for compile failure explicitly.

The code currently requires a compilation failure only if there is an error comment in the file. Note, this isn't currently useful other than ensuring that run-pass tests don't require that errors happen. And in that sense, it's a short-circuit, since it would probably result in the match failing below. Though, it does catch the unlikely failure mode where the compiler reports an error but doesn't return a failure code.